### DISCUSSION TITLE
What is the best way to import custom c++ library in Avalonia for different platforms?

### DISCUSSION THREAD

**Discussion by cangunaydin** [AvaloniaUI/Avalonia#13981] [Q&A]
Hello, 
I have custom c++ library that i want to invoke from c# code. I want to build the avalonia app for cross platform. 
As I know in ios you need to build your library as static. Correct me if i am wrong. is this the same for Avalonia app? It would be very helpful if there is a sample project or some doc that i look into it.
For ex in maui app, in ios platform i need to use <nativereference> tag 
![Screenshot 2023-12-17 at 20 12 49](https://github.com/AvaloniaUI/Avalonia/assets/3688369/c1fff6eb-1546-46ec-a754-212a4ee74935)

and use the dllimport with __Internal to invoke c++ functions. 
![Screenshot 2023-12-17 at 20 13 52](https://github.com/AvaloniaUI/Avalonia/assets/3688369/e864a457-e1e5-46e7-a8a0-6d514d9351d4)

should i do sth similar? I have tried to run the things with shared library as a normal import, and it seems like it works on ios simulator. So i am confused. I wonder if i can work with shared library on ios target, or should i build my library as static and use __Internal like maui app?

**Comment by timunie** [comment]
Maybe you can also use a soruce generator like MicroCom (check Avalonia soruce) to make it more trimming friendly. Otherwise this should work as shown above.

**Reply by cangunaydin** [reply]
so shared library is not an option? cause if i use custom import resolver it seems like it works on ios. or is this works only on iossimulator?
https://learn.microsoft.com/en-us/dotnet/standard/native-interop/native-library-loading#custom-import-resolver

**Reply by timunie** [reply]
I've no iOS nor a mac but I guess one needs to have a real world device. Using a simulator on Windows is not allowed (not our fault)

**Reply by cangunaydin** [reply]
ok, i think it is gonna be try, fail and find out for me in this case :) i will build a minimal shared lib for ios then i will try it on device and post it over here. thank you anyway.

**Comment by maxkatz6** [comment]
In general, any .NET mobile rules can also be applied to Avalonia, since Avalonia does not invent any custom mobile SDKs or workloads (yet?). Since MAUI works on the same .NET mobile SDKs, same rules can be reapplied. https://learn.microsoft.com/en-us/xamarin/ios/platform/native-interop

In general it seems like Apple iOS has some weird rules about using dylib in iOS apps, as it seems to have to be used as part of the framework packages. Which has less convenience in .NET.

**Comment by llfab** [comment]
For xplat development I propose ```LibraryImport``` over ```DllImport``` and for potentially different native lib names use ```NativeLibrary.SetDllImportResolver```
